### DISCUSSION TITLE
aiNodes

### DISCUSSION THREAD

**Discussion by XmYx** [bhowiebkr/python-node-editor#4] [Ideas]
Hi! I have just found out about your repo. I have been building a node app based on https://gitlab.com/pavel.krupala/pyqt-node-editor by Pavel Krupala, and it has been a great journey so far. If you feel like it, take a look at the project, we could join efforts if that suits you.
Admittedly, my engine is designed to run Ai computation, however, it wouldn't be far fetched to separate backend and frontend logic.
I'm eager to implement subgraphs, and handle their computation as well.

Here is the the repo mentioned:

https://github.com/XmYx/ainodes-engine

**Comment by bhowiebkr** [comment]
It looks very cool, and I’ll have to read through it and see if things can be combined in some way. 

I’m currently working on the logic side of this code base in a separate branch. Not sure if I’ll bring it into the main branch or build a separate repo for the code yet.

The idea I’m thinking is to pull out the node classes/types into separate files with a json scene description. Those custom nodes will override the functionality of the base class.

I really like how Unreal Engine does it with the blueprinting system and I think I’ll replicate some of its functionality in regards to flow execution. Doing so allows for loops, if statements and other datatype function nodes to enable some node based programming.

**Comment by XmYx** [comment]
Logic is pretty much how you thought - with the difference, that the serialization is not utilised, but each node has their own function, and makes a Worker QRunnable, runs it in the parents threadpool, resutl signal is connected to an onWorkerFinished function which takes care of gui outputs and runs the next child if any.

Now, what I'm also working on, is that I want to separate the execution from the frontend with a light FastApi server, also allowing processes to be ran on a server.

Same page on execution, so far I kept it as a separate input/output type, which only does execution, data is evaluated from the node if it is already present, and the execution flow is completely up to the user, no guidance in that (i.e. a graph with missing values in the beginning won't solve itself when evaluating a node in the end, but it, as you mentioned, allows loops, and data storage).

An other great repo I highly suggest to take a look at is NodeGraphQt. I took the inspiration for the backround node from that, and it works really great. It also has logic to handle group nodes, which are subgraphs practically, and that would also make life a lot easier.

It would be nice to find the defacto middleground, so we can provide a nice Qt node gui similar to chainner, but make it a lot more modular. (Optional if data and params are on the nodes or on a separate / docked widget, etc).

I'd love to hear your thoughts. Do you use discord maybe?

**Reply by bhowiebkr** [reply]
I'm on discord you can find me at Howie#4450

**Reply by XmYx** [reply]
Couldn't find you using that tag unfortunately. MIne is: miklos.mnagy#3822

**Reply by bhowiebkr** [reply]
ohh, yeah I changed it recently. It's: Bryan Howard#4450
